query/aggregate.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2015 Square Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package query
+
+// This file culminates in the definition of `aggregateBy`, which takes a SeriesList and an Aggregator and a list of tags,
+// and produces an aggregated SeriesList with one list per group, each group having been aggregated into it.
+
+import (
+	"math"
+
+	"github.com/square/metrics/api"
+)
+
+type group struct {
+	List   []api.Timeseries
+	TagSet api.TagSet
+}
+
+type groupResult struct {
+	Results []group
+}
+
+// If the given group will accept this given series (since it belongs to this group)
+// then bucketValid will return true.
+func bucketValid(left api.TagSet, right api.TagSet, tags []string) bool {
```

Comment:
Should standardize on `group` instead of `bucket`. This is more of a compatibility check between tags, so I think we'd benefit from naming this something along those lines.